README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

<div>
  <h2>Shimmer effect</h2>
  <p>The Shimmer effect package provides a customizable widget that adds a shimmering effect to any Flutter widget. This can be useful for indicating that a widget is currently loading or animating, and can be applied to buttons, text fields, images, and more.</p>
  <h3>Features</h3>
  <ul>
    <li>Easily add shimmering effect to any Flutter widget</li>
    <li>Customize the duration, color, and secondary color of the shimmer effect</li>
    <li>Support for multiple directions of shimmering effect</li>
  </ul>
  <h3>Getting Started</h3>
  <p>To use the Shimmer effect package, you'll need to add it to your pubspec.yaml file and import it into your Dart code:</p>
  <pre><code>dependencies:
    shimmer_effect: ^1.0.0
import 'package:shimmer_effect/shimmer_effect.dart';</code></pre>
  <h3>Usage</h3>
  <p>To add a shimmer effect to a Flutter widget, simply wrap it with the ShimmerEffect widget and specify the desired properties:</p>
  <pre><code>ShimmerEffect(
    duration: const Duration(seconds: 1),
    color: const Color.fromARGB(141, 240, 235, 235),
    secondaryColor: Colors.white,
    child: SizedBox(
      width: 200,
      height: 200,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: const [
          Icon(
            Icons.favorite,
            size: 72,
            color: Colors.red,
          ),
          SizedBox(height: 16),
          Text(
            'Shimmer effect',
            textAlign: TextAlign.center,
            style: TextStyle(fontSize: 24),
          ),
        ],
      ),
    ),
  ),</code></pre>
  <p>This code applies a shimmer effect to a container with a grey background and rounded corners, containing the text "Hello, world!" The shimmer effect will last for 2 seconds and use white and grey colors.</p>
  <p>For more examples, see the example directory in the Shimmer effect package on GitHub.</p>
</div>

CHANGELOG.md
## [1.0.1] - 2022-03-07
 
* Bug Fix.


## [1.0.0] - 2022-03-07
 
* Initial Release of shimmer effect.


test/shimmer_effect_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:shimmer_effect/shimmer_effect.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/shimmer_effect.dart
library shimmer_effect;

import 'package:flutter/material.dart';

class ShimmerEffect extends StatefulWidget {
  final Widget child;
  final Duration duration;
  final Color color;
  final Color secondaryColor;

  const ShimmerEffect({
    Key? key,
    required this.child,
    this.duration = const Duration(milliseconds: 1500),
    this.color = Colors.white,
    this.secondaryColor = Colors.grey,
  }) : super(key: key);

  @override
  // ignore: library_private_types_in_public_api
  _ShimmerEffectState createState() => _ShimmerEffectState();
}

class _ShimmerEffectState extends State<ShimmerEffect>
    with SingleTickerProviderStateMixin {
  late final AnimationController _controller;
  late final Animation<double> _animation;

  @override
  void initState() {
    super.initState();

    _controller = AnimationController(vsync: this, duration: widget.duration);

    _animation = Tween<double>(begin: -1.0, end: 2.0).animate(CurvedAnimation(
      parent: _controller,
      curve: Curves.linear,
      reverseCurve: Curves.linear,
    ));

    _controller.repeat();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animation,
      builder: (context, child) {
        return DecoratedBox(
          decoration: BoxDecoration(
            gradient: LinearGradient(
              colors: [
                widget.secondaryColor,
                widget.color,
                widget.secondaryColor,
              ],
              stops: [
                _animation.value + 0.3,
                _animation.value + 0.5,
                _animation.value + 0.7
              ],
              begin: const Alignment(-1, -1),
              end: const Alignment(1, 1),
              tileMode: TileMode.clamp,
            ),
          ),
          child: child,
        );
      },
      child: widget.child,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:shimmer_effect/shimmer_effect.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Shimmer Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Shimmer Animation Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        // backgroundColor: Colors.black,
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
          child: ShimmerEffect(
            duration: const Duration(seconds: 1),
            color: const Color.fromARGB(141, 240, 235, 235),
            secondaryColor: Colors.white,
            child: SizedBox(
              width: 200,
              height: 200,
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: const [
                  Icon(
                    Icons.favorite,
                    size: 72,
                    color: Colors.red,
                  ),
                  SizedBox(height: 16),
                  Text(
                    'Shimmer Animation',
                    textAlign: TextAlign.center,
                    style: TextStyle(fontSize: 24),
                  ),
                ],
              ),
            ),
          ),
        ));
  }
}
